Sources/NIO/BaseSocketChannel.swift

```diff
@@ -205,19 +205,84 @@ private struct SocketChannelLifecycleManager {
-    // Visible to access from EventLoop directly
+    struct AddressCache {
+        // deliberately lets because they must always be updated together (so forcing `init` is useful).
+        let local: Optional<SocketAddress>
+        let remote: Optional<SocketAddress>
+
+        init(local: SocketAddress?, remote: SocketAddress?) {
+            self.local = local
+            self.remote = remote
+        }
+    }
+
+    // MARK: - Stored Properties
+    // MARK: Constants & atomics (accessible everywhere)
     public let parent: Channel?
     internal let socket: SocketType
     private let closePromise: EventLoopPromise<Void>
     internal let selectableEventLoop: SelectableEventLoop
-    private let addressesCached: AtomicBox<Box<(local:SocketAddress?, remote:SocketAddress?)>> = AtomicBox(value: Box((local: nil, remote: nil)))
-    private let bufferAllocatorCached: AtomicBox<Box<ByteBufferAllocator>>
+    private let _offEventLoopLock = Lock()
     private let isActiveAtomic: NIOAtomic<Bool> = .makeAtomic(value: false)
-    private var _pipeline: ChannelPipeline! = nil // this is really a constant (set in .init) but needs `self` to be constructed and therefore a `var`. Do not change as this needs to accessed from arbitrary threads
     // just a thread-safe way of having something to print about the socket from any thread
     internal let socketDescription: String
 
+    // MARK: Variables, on EventLoop thread only
+    var readPending = false
+    var pendingConnect: Optional<EventLoopPromise<Void>>
+    var recvAllocator: RecvByteBufferAllocator
+    var maxMessagesPerRead: UInt = 4
+    private var inFlushNow: Bool = false // Guard against re-entrance of flushNow() method.
+    private var autoRead: Bool = true
+
+    // MARK: Varibles that are really constants
+    private var _pipeline: ChannelPipeline! = nil // this is really a constant (set in .init) but needs `self` to be constructed and therefore a `var`. Do not change as this needs to accessed from arbitrary threads
+
+    // MARK: Special variables, please read comments.
+    // For reads guarded by _offEventLoopLock/EL thread; writes guarded by _offEventLoopLock (and on EL thread only).
+    // PLEASE don't use these directly and use the non-underscored computed properties instead.
+    private var _addressCache = AddressCache(local: nil, remote: nil) // please use `self.addressesCached` instead
+    private var _bufferAllocatorCache: ByteBufferAllocator  // please use `self.bufferAllocatorCached` instead.
+
+    // MARK: - Computed properties
+    // This is called from arbirary threads.
+    internal var addressesCached: AddressCache {
+        get {
+            if self.eventLoop.inEventLoop {
+                return self._addressCache
+            } else {
+                return self._offEventLoopLock.withLock {
+                    return self._addressCache
+                }
+            }
+        }
+        set {
+            assert(self.eventLoop.inEventLoop)
```

Comment:
Please use `assertInEventLoop`, but in this case given how important this is to the safety of the program I think this should be `preconditionInEventLoop`.